File: src/appearance.c
/* SPDX-License-Identifier: MIT
 *
 * Dark mode, via the desktop settings portal.
 *
 * A compositor cannot make its clients dark by styling anything: Firefox, GTK
 * and Qt apps each decide for themselves, and they all ask the same question —
 * the `color-scheme` key in the `org.freedesktop.appearance` namespace, read
 * over D-Bus through xdg-desktop-portal. Without something answering, every
 * application falls back to light regardless of how the shell looks.
 *
 * Normally that answer comes from a desktop environment. There isn't one here,
 * so the compositor provides it directly by implementing
 * org.freedesktop.impl.portal.Settings. This is deliberately not done by
 * setting GSettings keys: that route needs dconf and GNOME's schemas
 * installed, and silently does nothing when they are absent.
 *
 * Values follow the spec: 0 no preference, 1 prefer dark, 2 prefer light.
 */
#define _POSIX_C_SOURCE 200809L

#include <stdlib.h>
#include <string.h>

#include <gio/gio.h>

#include <wlr/util/log.h>

#include "viewport.h"

#define PORTAL_BUS_NAME "org.freedesktop.impl.portal.desktop.viewport"
#define PORTAL_OBJECT_PATH "/org/freedesktop/portal/desktop"
#define SETTINGS_INTERFACE "org.freedesktop.impl.portal.Settings"
#define APPEARANCE_NAMESPACE "org.freedesktop.appearance"
#define GNOME_NAMESPACE "org.gnome.desktop.interface"

struct viewport_appearance {
	struct viewport_server *server;
	GDBusConnection *connection;
	guint owner_id;
	guint registration_id;
	/* 1 = prefer dark, 2 = prefer light, 0 = no preference. */
	guint32 color_scheme;
};

static const char introspection_xml[] =
	"<node>"
	"  <interface name='" SETTINGS_INTERFACE "'>"
	"    <method name='ReadAll'>"
	"      <arg type='as' name='namespaces' direction='in'/>"
	"      <arg type='a{sa{sv}}' name='value' direction='out'/>"
	"    </method>"
	"    <method name='Read'>"
	"      <arg type='s' name='namespace' direction='in'/>"
	"      <arg type='s' name='key' direction='in'/>"
	"      <arg type='v' name='value' direction='out'/>"
	"    </method>"
	"    <signal name='SettingChanged'>"
	"      <arg type='s' name='namespace'/>"
	"      <arg type='s' name='key'/>"
	"      <arg type='v' name='value'/>"
	"    </signal>"
	"    <property name='version' type='u' access='read'/>"
	"  </interface>"
	"</node>";

static GDBusNodeInfo *introspection_data;

/* Font and scaling settings.
 *
 * Answering only color-scheme is worse than providing no portal at all: a
 * client that finds a working Settings implementation trusts it instead of
 * falling back to its own defaults, so a missing font or text-scaling value
 * becomes a measurement of zero rather than a sensible default. Toolkits size
 * menus from exactly these keys.
 *
 * The values are conventional defaults rather than anything clever — the point
 * is that every key a toolkit asks for gets a usable answer. */
static void add_gnome_settings(struct viewport_appearance *appearance,
	GVariantBuilder *builder)
{
	g_variant_builder_add(builder, "{sv}", "color-scheme",
		g_variant_new_string(appearance->color_scheme == 1
			? "prefer-dark" : "prefer-light"));
	g_variant_builder_add(builder, "{sv}", "gtk-theme",
		g_variant_new_string(appearance->color_scheme == 1
			? "Adwaita-dark" : "Adwaita"));
	g_variant_builder_add(builder, "{sv}", "icon-theme",
		g_variant_new_string("Adwaita"));
	g_variant_builder_add(builder, "{sv}", "font-name",
		g_variant_new_string("Sans 10"));
	g_variant_builder_add(builder, "{sv}", "monospace-font-name",
		g_variant_new_string("Monospace 10"));
	g_variant_builder_add(builder, "{sv}", "document-font-name",
		g_variant_new_string("Sans 10"));
	g_variant_builder_add(builder, "{sv}", "text-scaling-factor",
		g_variant_new_double(1.0));
	g_variant_builder_add(builder, "{sv}", "cursor-theme",
		g_variant_new_string("default"));
	g_variant_builder_add(builder, "{sv}", "cursor-size",
		g_variant_new_int32(24));
	g_variant_builder_add(builder, "{sv}", "enable-animations",
		g_variant_new_boolean(TRUE));
}

/* Returns a floating value, or NULL when the key is genuinely unknown. */
static GVariant *lookup_setting(struct viewport_appearance *appearance,
	const char *namespace, const char *key)
{
	if (strcmp(namespace, APPEARANCE_NAMESPACE) == 0) {
		if (strcmp(key, "color-scheme") == 0) {
			return g_variant_new_uint32(appearance->color_scheme);
		}
		/* Documented in the appearance namespace; clients read both. */
		if (strcmp(key, "contrast") == 0) {
			return g_variant_new_uint32(0);
		}
		return NULL;
	}

	if (strcmp(namespace, GNOME_NAMESPACE) != 0) {
		return NULL;
	}

	GVariantBuilder builder;
	g_variant_builder_init(&builder, G_VARIANT_TYPE("a{sv}"));
	add_gnome_settings(appearance, &builder);
	GVariant *all = g_variant_ref_sink(g_variant_builder_end(&builder));

	GVariant *value = g_variant_lookup_value(all, key, NULL);
	g_variant_unref(all);
	return value;
}

static void handle_method(GDBusConnection *connection, const char *sender,
	const char *object_path, const char *interface_name,
	const char *method_name, GVariant *parameters,
	GDBusMethodInvocation *invocation, gpointer user_data)
{
	struct viewport_appearance *appearance = user_data;

	if (strcmp(method_name, "ReadAll") == 0) {
		GVariantBuilder namespaces;
		g_variant_builder_init(&namespaces, G_VARIANT_TYPE("a{sa{sv}}"));

		GVariantBuilder appearance_keys;
		g_variant_builder_init(&appearance_keys, G_VARIANT_TYPE("a{sv}"));
		g_variant_builder_add(&appearance_keys, "{sv}", "color-scheme",
			g_variant_new_uint32(appearance->color_scheme));
		g_variant_builder_add(&namespaces, "{sa{sv}}", APPEARANCE_NAMESPACE,
			&appearance_keys);

		GVariantBuilder gnome_keys;
		g_variant_builder_init(&gnome_keys, G_VARIANT_TYPE("a{sv}"));
		add_gnome_settings(appearance, &gnome_keys);
		g_variant_builder_add(&namespaces, "{sa{sv}}", GNOME_NAMESPACE,
			&gnome_keys);

		g_dbus_method_invocation_return_value(invocation,
			g_variant_new("(a{sa{sv}})", &namespaces));
		return;
	}

	if (strcmp(method_name, "Read") == 0) {
		const char *namespace = NULL, *key = NULL;
		g_variant_get(parameters, "(&s&s)", &namespace, &key);

		GVariant *value = lookup_setting(appearance, namespace, key);
		if (value != NULL) {
			g_dbus_method_invocation_return_value(invocation,
				g_variant_new("(v)", value));
			return;
		}

		/* The spec requires this specific error for anything unknown, and
		 * clients rely on it to fall back cleanly. */
		g_dbus_method_invocation_return_error(invocation, G_DBUS_ERROR,
			G_DBUS_ERROR_UNKNOWN_PROPERTY, "unknown setting %s/%s", namespace,
			key);
		return;
	}

	g_dbus_method_invocation_return_error(invocation, G_DBUS_ERROR,
		G_DBUS_ERROR_UNKNOWN_METHOD, "unknown method %s", method_name);
}

static GVariant *handle_get_property(GDBusConnection *connection,
	const char *sender, const char *object_path, const char *interface_name,
	const char *property_name, GError **error, gpointer user_data)
{
	if (strcmp(property_name, "version") == 0) {
		return g_variant_new_uint32(2);
	}
	return NULL;
}

static const GDBusInterfaceVTable interface_vtable = {
	.method_call = handle_method,
	.get_property = handle_get_property,
};

static void handle_bus_acquired(GDBusConnection *connection, const char *name,
	gpointer user_data)
{
	struct viewport_appearance *appearance = user_data;
	GError *error = NULL;

	appearance->connection = connection;
	appearance->registration_id = g_dbus_connection_register_object(connection,
		PORTAL_OBJECT_PATH, introspection_data->interfaces[0],
		&interface_vtable, appearance, NULL, &error);

	if (appearance->registration_id == 0) {
		wlr_log(WLR_ERROR, "settings portal: %s",
			error ? error->message : "registration failed");
		g_clear_error(&error);
		return;
	}

	wlr_log(WLR_INFO, "settings portal up, color-scheme=%u",
		appearance->color_scheme);
}

static void handle_name_lost(GDBusConnection *connection, const char *name,
	gpointer user_data)
{
	/* Not fatal: without a session bus, or with another portal already owning
	 * the name, applications simply keep their own defaults. */
	wlr_log(WLR_INFO,
		"settings portal unavailable; clients will use their own defaults");
}

void viewport_appearance_set_dark(struct viewport_appearance *appearance,
	bool dark)
{
	if (appearance == NULL) {
		return;
	}

	guint32 value = dark ? 1 : 2;
	if (appearance->color_scheme == value) {
		return;
	}
	appearance->color_scheme = value;

	if (appearance->connection == NULL) {
		return;
	}

	/* Running applications switch on this signal rather than at next start. */
	g_dbus_connection_emit_signal(appearance->connection, NULL,
		PORTAL_OBJECT_PATH, SETTINGS_INTERFACE, "SettingChanged",
		g_variant_new("(ssv)", APPEARANCE_NAMESPACE, "color-scheme",
			g_variant_new_uint32(value)),
		NULL);

	wlr_log(WLR_INFO, "color-scheme now %s", dark ? "dark" : "light");
}

bool viewport_appearance_is_dark(struct viewport_appearance *appearance)
{
	return appearance != NULL && appearance->color_scheme == 1;
}

struct viewport_appearance *viewport_appearance_create(
	struct viewport_server *server, bool dark)
{
	struct viewport_appearance *appearance = calloc(1, sizeof(*appearance));
	if (appearance == NULL) {
		return NULL;
	}
	appearance->server = server;
	appearance->color_scheme = dark ? 1 : 2;

	GError *error = NULL;
	if (introspection_data == NULL) {
		introspection_data = g_dbus_node_info_new_for_xml(introspection_xml,
			&error);
		if (introspection_data == NULL) {
			wlr_log(WLR_ERROR, "settings portal introspection: %s",
				error ? error->message : "parse failed");
			g_clear_error(&error);
			free(appearance);
			return NULL;
		}
	}

	/* G_BUS_NAME_OWNER_FLAGS_REPLACE loses to a real desktop portal rather
	 * than fighting it, which matters if this is ever run inside a session
	 * that already provides one. */
	appearance->owner_id = g_bus_own_name(G_BUS_TYPE_SESSION, PORTAL_BUS_NAME,
		G_BUS_NAME_OWNER_FLAGS_NONE, handle_bus_acquired, NULL,
		handle_name_lost, appearance, NULL);

	return appearance;
}

void viewport_appearance_destroy(struct viewport_appearance *appearance)
{
	if (appearance == NULL) {
		return;
	}
	if (appearance->registration_id != 0 && appearance->connection != NULL) {
		g_dbus_connection_unregister_object(appearance->connection,
			appearance->registration_id);
	}
	if (appearance->owner_id != 0) {
		g_bus_unown_name(appearance->owner_id);
	}
	free(appearance);
}

File: crates/viewport/src/appearance.rs
// SPDX-License-Identifier: GPL-3.0-or-later
//
// Dark mode, via the desktop settings portal. Ports src/appearance.c.
//
// A compositor cannot make its clients dark by styling anything: Firefox, GTK
// and Qt apps each decide for themselves, and they all ask the same question —
// the `color-scheme` key in the `org.freedesktop.appearance` namespace, read
// over D-Bus through xdg-desktop-portal. With nothing answering, every
// application falls back to light however the shell looks.
//
// Normally that answer comes from a desktop environment. There is not one here,
// so the compositor answers directly by implementing
// org.freedesktop.impl.portal.Settings. Deliberately not by setting GSettings
// keys: that route needs dconf and GNOME's schemas installed, and silently does
// nothing when they are absent.
//
// It shares the D-Bus thread arrangement with `notification.rs` — zbus wants an
// async runtime and this loop is GLib with calloop inside it — but not its
// channel: everything here is answered from state the object already holds, so
// nothing has to reach the compositor to reply.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use zvariant::{OwnedValue, Value};

const BUS_NAME: &str = "org.freedesktop.impl.portal.desktop.viewport";
const OBJECT_PATH: &str = "/org/freedesktop/portal/desktop";
const INTERFACE: &str = "org.freedesktop.impl.portal.Settings";
const APPEARANCE: &str = "org.freedesktop.appearance";
const GNOME: &str = "org.gnome.desktop.interface";

/// `color-scheme` as the specification numbers it.
///
/// The numbers are the interface: a client reads the integer, so 1 has to mean
/// dark whatever this compositor would rather call it.
pub const PREFER_DARK: u32 = 1;
pub const PREFER_LIGHT: u32 = 2;

/// What the portal answers with.
#[derive(Debug, Clone)]
pub struct Settings {
    pub color_scheme: u32,
    /// The cursor the compositor actually draws. A toolkit sizes its own
    /// cursors from these, and a disagreement here is a cursor that changes
    /// size as it crosses from a window into the desktop.
    pub cursor_theme: String,
    pub cursor_size: i32,
}

impl Default for Settings {
    fn default() -> Self {
        Self {
            color_scheme: PREFER_DARK,
            cursor_theme: "default".to_owned(),
            cursor_size: 24,
        }
    }
}

/// A value the way D-Bus carries it.
///
/// Infallible for everything here: only file descriptors can fail to be owned,
/// and a setting is never one.
fn owned<'a, T: Into<Value<'a>>>(value: T) -> OwnedValue {
    value
        .into()
        .try_to_owned()
        .expect("a setting is never a file descriptor")
}

/// The `org.gnome.desktop.interface` keys.
///
/// Answering only color-scheme is worse than providing no portal at all: a
/// client that finds a working Settings implementation trusts it instead of
/// falling back to its own defaults, so a missing font or text-scaling value
/// becomes a measurement of zero rather than a sensible default. Toolkits size
/// menus from exactly these keys.
///
/// The values are conventional defaults rather than anything clever — the point
/// is that every key a toolkit asks for gets a usable answer.
pub fn gnome_settings(settings: &Settings) -> HashMap<String, OwnedValue> {
    let dark = settings.color_scheme == PREFER_DARK;
    let mut keys = HashMap::new();
    keys.insert(
        "color-scheme".to_owned(),
        owned(if dark { "prefer-dark" } else { "prefer-light" }),
    );
    keys.insert(
        "gtk-theme".to_owned(),
        owned(if dark { "Adwaita-dark" } else { "Adwaita" }),
    );
    keys.insert("icon-theme".to_owned(), owned("Adwaita"));
    keys.insert("font-name".to_owned(), owned("Sans 10"));
    keys.insert("monospace-font-name".to_owned(), owned("Monospace 10"));
    keys.insert("document-font-name".to_owned(), owned("Sans 10"));
    keys.insert("text-scaling-factor".to_owned(), owned(1.0f64));
    keys.insert(
        "cursor-theme".to_owned(),
        owned(settings.cursor_theme.as_str()),
    );
    keys.insert("cursor-size".to_owned(), owned(settings.cursor_size));
    keys.insert("enable-animations".to_owned(), owned(true));
    keys
}

/// The `org.freedesktop.appearance` keys.
pub fn appearance_settings(settings: &Settings) -> HashMap<String, OwnedValue> {
    let mut keys = HashMap::new();
    keys.insert("color-scheme".to_owned(), owned(settings.color_scheme));
    // Documented in the same namespace, and clients read both.
    keys.insert("contrast".to_owned(), owned(0u32));
    keys
}

/// One setting, or `None` when the key is genuinely unknown.
///
/// Unknown has to stay distinguishable from absent: the specification requires
/// an error for a key that does not exist, and clients fall back cleanly on it
/// where they would take a zero at face value.
pub fn lookup(settings: &Settings, namespace: &str, key: &str) -> Option<OwnedValue> {
    match namespace {
        APPEARANCE => appearance_settings(settings).remove(key),
        GNOME => gnome_settings(settings).remove(key),
        _ => None,
    }
}

/// Every namespace, as `ReadAll` returns them.
///
/// The requested patterns are ignored, as in `src/appearance.c:151`: a client
/// asking for one namespace and being given two has what it asked for, and
/// matching the specification's glob syntax here would be a parser whose only
/// effect is to withhold answers.
pub fn read_all(settings: &Settings) -> HashMap<String, HashMap<String, OwnedValue>> {
    HashMap::from([
        (APPEARANCE.to_owned(), appearance_settings(settings)),
        (GNOME.to_owned(), gnome_settings(settings)),
    ])
}

/// The half the compositor keeps.
#[derive(Default)]
pub struct Appearance {
    settings: Arc<Mutex<Settings>>,
    /// Kept to emit `SettingChanged`, and `None` when the name could not be
    /// claimed.
    connection: Option<zbus::blocking::Connection>,
}

impl Appearance {
    /// Claim the name and start answering.
    ///
    /// A failure is not fatal: a session with no D-Bus, or one where a real
    /// desktop portal already owns the name, still has a working compositor —
    /// its applications simply keep their own defaults, which is what they had
    /// a moment ago.
    pub fn start(
        &mut self,
        settings: Settings,
        screencast: crate::screencast::portal::ScreenCast,
    ) -> anyhow::Result<()> {
        let scheme = settings.color_scheme;
        *self.settings.lock().unwrap() = settings;

        let portal = Portal {
            settings: self.settings.clone(),
        };
        // One connection for both interfaces, because they share a bus name.
        //
        // A second connection claiming org.freedesktop.impl.portal.desktop.
        // viewport does not get it — the first one holds it — so whichever
        // interface was built second was simply absent from the bus, and the
        // portal frontend fell through to another backend without saying
        // anything. Settings and ScreenCast live at the same object path, as
        // they do in every other portal implementation.
        //
        // The name is claimed if it is free and left alone if it is not,
        // rather than replacing whoever holds it: a real desktop portal
        // running alongside knows more about the session than this does.
        // Before it is handed over: the watcher needs the same sessions the
        // object on the bus is keeping.
        let sessions = screencast.sessions();
        let closer = screencast.closer();

        let connection = zbus::blocking::connection::Builder::session()?
            .name(BUS_NAME)?
            .serve_at(OBJECT_PATH, portal)?
            .serve_at(OBJECT_PATH, screencast)?
            .build()?;

        crate::screencast::portal::watch_frontend(connection.clone(), sessions, closer);

        self.connection = Some(connection);
        tracing::info!("settings and screencast portals up, color-scheme={scheme}");
        Ok(())
    }

    pub fn is_dark(&self) -> bool {
        self.settings.lock().unwrap().color_scheme == PREFER_DARK
    }

    /// Switch the session between dark and light.
    ///
    /// Running applications change on the signal rather than at next start,
    /// which is the whole reason the portal is a service and not a file.
    pub fn set_dark(&mut self, dark: bool) {
        let value = if dark { PREFER_DARK } else { PREFER_LIGHT };
        {
            let mut settings = self.settings.lock().unwrap();
            if settings.color_scheme == value {
                return;
            }
            settings.color_scheme = value;
        }

        let Some(connection) = self.connection.as_ref() else {
            return;
        };
        if let Err(e) = connection.emit_signal(
            None::<&str>,
            OBJECT_PATH,
            INTERFACE,
            "SettingChanged",
            &(APPEARANCE, "color-scheme", Value::from(value)),
        ) {
            tracing::warn!("could not announce the colour scheme: {e}");
        }
        tracing::info!("color-scheme now {}", if dark { "dark" } else { "light" });
    }
}

/// The object on the bus.
struct Portal {
    settings: Arc<Mutex<Settings>>,
}

#[zbus::interface(name = "org.freedesktop.impl.portal.Settings")]
impl Portal {
    fn read_all(&self, _namespaces: Vec<String>) -> HashMap<String, HashMap<String, OwnedValue>> {
        read_all(&self.settings.lock().unwrap())
    }

    fn read(&self, namespace: &str, key: &str) -> zbus::fdo::Result<OwnedValue> {
        lookup(&self.settings.lock().unwrap(), namespace, key).ok_or_else(|| {
            // This specific error, because it is what a client checks for
            // before falling back to its own default.
            zbus::fdo::Error::UnknownProperty(format!("unknown setting {namespace}/{key}"))
        })
    }

    /// Announced when the scheme changes, so a running application switches
    /// rather than waiting to be restarted.
    ///
    /// Declared here only to put it in the introspection — it is emitted
    /// through the connection, from `set_dark`, where the compositor knows the
    /// scheme changed. A client subscribes by match rule either way.
    #[zbus(signal)]
    async fn setting_changed(
        emitter: &zbus::object_server::SignalEmitter<'_>,
        namespace: &str,
        key: &str,
        value: Value<'_>,
    ) -> zbus::Result<()>;

    // Lower case, and named explicitly because zbus would otherwise publish it
    // as `Version` — the interface spells it `version`, and a client reading
    // the version to decide what it may ask for finds nothing at all.
    #[zbus(property, name = "version")]
    fn version(&self) -> u32 {
        2
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn dark() -> Settings {
        Settings {
            color_scheme: PREFER_DARK,
            cursor_theme: "Bibata".to_owned(),
            cursor_size: 32,
        }
    }

    #[test]
    fn the_scheme_is_the_number_clients_read() {
        // The integer is the interface. Naming them the other way round makes
        // every application light while the shell is dark, and nothing in the
        // log says so.
        let keys = appearance_settings(&dark());
        assert_eq!(u32::try_from(&keys["color-scheme"]).unwrap(), 1);
        let light = Settings {
            color_scheme: PREFER_LIGHT,
            ..dark()
        };
        assert_eq!(
            u32::try_from(&appearance_settings(&light)["color-scheme"]).unwrap(),
            2
        );
    }

    #[test]
    fn the_gnome_namespace_says_the_same_thing_in_its_own_words() {
        // GTK reads the string here rather than the integer above, so the two
        // have to agree — a client that got them from different sources would
        // draw a dark window with a light titlebar.
        let keys = gnome_settings(&dark());
        assert_eq!(
            String::try_from(keys["color-scheme"].clone()).unwrap(),
            "prefer-dark"
        );
        assert_eq!(
            String::try_from(keys["gtk-theme"].clone()).unwrap(),
            "Adwaita-dark"
        );
    }

    #[test]
    fn every_key_a_toolkit_asks_for_has_an_answer() {
        // A client that finds a working portal trusts it instead of falling
        // back, so a missing font becomes a measurement of zero rather than a
        // sensible default.
        let keys = gnome_settings(&dark());
        for key in [
            "font-name",
            "monospace-font-name",
            "document-font-name",
            "text-scaling-factor",
            "icon-theme",
            "cursor-theme",
            "cursor-size",
            "enable-animations",
        ] {
            assert!(keys.contains_key(key), "{key} has no answer");
        }
    }

    #[test]
    fn the_cursor_reported_is_the_cursor_drawn() {
        // A toolkit sizes its own cursors from these. Reporting a default while
        // the compositor draws something else is a pointer that changes size as
        // it crosses from a window onto the desktop.
        let keys = gnome_settings(&dark());
        assert_eq!(
            String::try_from(keys["cursor-theme"].clone()).unwrap(),
            "Bibata"
        );
        assert_eq!(i32::try_from(&keys["cursor-size"]).unwrap(), 32);
    }

    #[test]
    fn an_unknown_key_is_unknown_rather_than_zero() {
        // The specification requires an error, and clients fall back cleanly on
        // it where they would take a zero at face value.
        assert!(lookup(&dark(), APPEARANCE, "colour-scheme").is_none());
        assert!(lookup(&dark(), "com.example.whatever", "color-scheme").is_none());
        assert!(lookup(&dark(), APPEARANCE, "color-scheme").is_some());
        assert!(lookup(&dark(), GNOME, "font-name").is_some());
    }

    #[test]
    fn read_all_carries_both_namespaces() {
        let all = read_all(&dark());
        assert!(all.contains_key(APPEARANCE));
        assert!(all.contains_key(GNOME));
    }
}
